File: src/Total.Recall.Scanners.Python/src/total_recall_scan/jsonl.py
"""JSONL writer matching the Total.Recall canonical contract.

Per docs/SCANNER_SCHEMA.md:
- One JSON object per line.
- UTF-8, no BOM.
- NO trailing newline after the last record.
- Compact separators (no insignificant whitespace) — matches the .NET writer.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Iterable


def write_jsonl(path: str | Path, records: Iterable[dict[str, Any]]) -> int:
    """Write ``records`` to ``path`` as canonical JSONL. Returns count written.

    Overwrites the target file. Records are serialised with ``ensure_ascii=False``
    so non-ASCII identifiers (Python supports them) survive round-trip, and with
    compact separators to match the .NET ``System.Text.Json`` default.
    """
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)

    materialised = list(records)
    if not materialised:
        # Per spec: empty store is a zero-byte file, not absent.
        path.write_bytes(b"")
        return 0

    lines = [
        json.dumps(rec, ensure_ascii=False, separators=(",", ":"), sort_keys=False)
        for rec in materialised
    ]
    # No trailing newline — the .NET writer omits it on the final record.
    path.write_text("\n".join(lines), encoding="utf-8", newline="")
    return len(lines)

File: src/Total.Recall.Scanners.TypeScript/src/jsonl.ts
/**
 * JSONL writer — byte-identical to the .NET and Python scanners.
 *
 * Compact JSON (no spaces), one record per line, joined with "\n",
 * no trailing newline. Empty input produces a zero-byte file.
 */

import { writeFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

export function writeJsonl(path: string, records: ReadonlyArray<unknown>): number {
  mkdirSync(dirname(path), { recursive: true });
  if (records.length === 0) {
    writeFileSync(path, "");
    return 0;
  }
  const body = records.map((r) => JSON.stringify(r)).join("\n");
  writeFileSync(path, body);
  return records.length;
}
